data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java

```diff
@@ -110,25 +124,37 @@ private WireEvent getWireEvent(final Event event) {
         );
     }
 
-    private CompletableFuture<AggregatedHttpResponse> processHttpRequest(final WebClient client, final String content) {
+    private CompletableFuture<AggregatedHttpResponse> processHttpRequest(final WebClient client, final String content,
+                                                                         final Collection<Record<Event>> records,
+                                                                         final PeerForwarderReceiveBuffer<Record<Event>> peerForwarderReceiveBuffer) {
         return CompletableFuture.supplyAsync(() ->
         {
-            final CompletableFuture<AggregatedHttpResponse> aggregate = client.post(DEFAULT_PEER_FORWARDING_URI, content).aggregate();
-            return aggregate.join();
-        }, executorService);
+            try {
+                final CompletableFuture<AggregatedHttpResponse> aggregate = client.post(DEFAULT_PEER_FORWARDING_URI, content).aggregate();
+                return aggregate.join();
+            } catch (final Exception e) {
+                processFailedRequestsLocally(null, records, peerForwarderReceiveBuffer);
+                throw e;
+            }
+        }, executorService).thenApply(httpResponse -> processFailedRequestsLocally(httpResponse, records, peerForwarderReceiveBuffer));
     }
 
-    private AggregatedHttpResponse getAggregatedHttpResponse(final CompletableFuture<AggregatedHttpResponse> aggregatedHttpResponseCompletableFuture) throws UnprocessedRequestException {
-        try {
-            return aggregatedHttpResponseCompletableFuture.get();
-        } catch (final InterruptedException e) {
-            LOG.error("Peer forwarding interrupted.");
-            throw new RuntimeException(e);
-        } catch (final ExecutionException e) {
-            if (e.getCause() instanceof RuntimeException)
-                throw (RuntimeException) e.getCause();
-            throw new RuntimeException(e.getCause());
+    private AggregatedHttpResponse processFailedRequestsLocally(final AggregatedHttpResponse httpResponse, final Collection<Record<Event>> records,
```

Comment:
From a class-design perspective, I don't think the `PeerForwarderClient` should be responsible for the logic of handling failed requests.

I would suggest that you move `processFailedRequestsLocally` to the `RemotePeerForwarder` class. Then the `PeerForwarderClient` simply receives a method (perhaps a `Consumer<AggregatedHttpResponse>`). Then in `RemotePeerForwarder`:

```
 peerForwarderClient.serializeRecordsAndSendHttpRequest(entry.getValue(),
                             destinationIp, pluginId, pipelineName, failedResponse -> processFailedRequestsLocally(failedResponse, entry.getValue(), peerForwarderReceiveBuffer);
```

You could even take the step of moving the `processFailedRequestsLocally` into a dedicated class for error handling. It might make unit testing easier.

Reply:
Fair point. Passing the buffer to the client felt wrong, I like this approach better